python/lsst/meas/astrom/pessimistic_pattern_matcher_b_3D.py

```diff
@@ -0,0 +1,1111 @@
+                          "This could end badly.")
+        center_vect = np.nanmean(source_array, axis=0)
+
+        # So that our rotation test works over the full sky we compute
+        # the max extent in each cartesian direction x,y,z.
+        xbtm_vect = np.array([np.min(source_array[:, 0]), center_vect[1],
+                              center_vect[2]], dtype=np.float64)
+        xtop_vect = np.array([np.max(source_array[:, 0]), center_vect[1],
+                              center_vect[2]], dtype=np.float64)
+        xbtm_vect /= np.sqrt(np.dot(xbtm_vect, xbtm_vect))
+        xtop_vect /= np.sqrt(np.dot(xtop_vect, xtop_vect))
+
+        ybtm_vect = np.array([center_vect[0], np.min(source_array[:, 1]),
+                              center_vect[2]], dtype=np.float64)
+        ytop_vect = np.array([center_vect[0], np.max(source_array[:, 1]),
+                              center_vect[2]], dtype=np.float64)
+        ybtm_vect /= np.sqrt(np.dot(ybtm_vect, ybtm_vect))
+        ytop_vect /= np.sqrt(np.dot(ytop_vect, ytop_vect))
+
+        zbtm_vect = np.array([center_vect[0], center_vect[1],
+                              np.min(source_array[:, 2])], dtype=np.float64)
+        ztop_vect = np.array([center_vect[0], center_vect[1],
+                              np.max(source_array[:, 2])], dtype=np.float64)
+        zbtm_vect /= np.sqrt(np.dot(zbtm_vect, zbtm_vect))
+        ztop_vect /= np.sqrt(np.dot(ztop_vect, ztop_vect))
+
+        # Return our list of vectors for later rotation testing.
+        return [xbtm_vect, xtop_vect, ybtm_vect, ytop_vect,
+                zbtm_vect, ztop_vect]
+
+    def _construct_pattern_and_shift_rot_matrix(self, src_pattern_array,
+                                                n_match, max_cos_theta_shift,
+                                                max_cos_rot_sq, max_dist_rad):
+        r"""Test an input source pattern against the reference catalog.
+        Returns the candidate matched patterns and their
+        implied rotation matrices or None.
+
+        Parameters
+        ----------
+        src_pattern_array : float array
+            Sub selection of source 3 vectors to create a pattern from
+        n_match : int
+            Number of points to attempt to create a pattern from. Must be
+            >= len(src_pattern_array)
+        max_cos_theta_shift : float
+            Maximum shift allowed between two patterns' centers.
+        max_cos_rot_sq : float
+            Maximum rotation beteween two patterns that have been shifted
+            to have their centers on top of each other.
+        max_dist_rad : float
+            Maximum delta distance allowed between the source and reference
+            pair distances to consider the reference pair a candidate for
+            the source pair. Also sets the tolerance between the opening
+            angles of the spokes when compared to the reference.
+
+        Returns
+        -------
+        None or lsst.pipe.base.Struct
```

Comment:
There are no cases where I really want to compute partial data. In the case of the matcher, return a partial result could actually be worse than returning nothing. I understand how this could work in other code, however I would rather have a hard failure that is easy to track than a soft one that can act like a success when in fact the results are extremely wrong. I've started using the structs as requested but on failures I set them to empty lists or None type as in most places I do not want a partial result.